internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubProducerAdapterConcurrentDelegator.java

```diff
@@ -0,0 +1,279 @@
+ *
+ * This pattern is mainly useful for stream processing as one container can write to multiple partitions.
+ */
+public class PubSubProducerAdapterConcurrentDelegator implements PubSubProducerAdapter {
+  private static final Logger LOGGER = LogManager.getLogger(PubSubProducerAdapterConcurrentDelegator.class);
+  private final String producingTopic;
+  private final ExecutorService producerExecutor;
+  private final List<MemoryBoundBlockingQueue<ProducerQueueNode>> blockingQueueList;
+  private final Map<MemoryBoundBlockingQueue<ProducerQueueNode>, ProducerQueueNode> lastNodePerQueue;
+  private final AtomicInteger selectedQueueId = new AtomicInteger(0);
+  private final Map<Integer, MemoryBoundBlockingQueue<ProducerQueueNode>> partitionQueueAssignment;
+  private final List<PubSubProducerAdapter> producers;
+  private final List<ProducerQueueDrainer> drainers;
+
+  public PubSubProducerAdapterConcurrentDelegator(
+      String producingTopic,
+      int producerThreadCount,
+      int producerQueueSize,
+      Supplier<PubSubProducerAdapter> producerAdapterSupplier) {
+    if (producerThreadCount <= 1) {
+      throw new VeniceException("'producerThreadCount' should be larger than 1, but got " + producerThreadCount);
+    }
+    if (producerQueueSize <= 2 * 1024 * 1024) {
+      throw new VeniceException("'producerQueueSize' should be larger than 2MB, but got " + producerQueueSize);
+    }
+    LOGGER.info(
+        "Initializing a 'PubSubProducerAdapterConcurrentDelegator' instance for topic: {} with thread"
+            + " count: {} and queue size: {}",
+        producingTopic,
+        producerThreadCount,
+        producerQueueSize);
+    this.blockingQueueList = new ArrayList<>(producerThreadCount);
+    this.producerExecutor = Executors
+        .newFixedThreadPool(producerThreadCount, new DaemonThreadFactory(producingTopic + "-concurrent-producer"));
+    this.partitionQueueAssignment = new VeniceConcurrentHashMap<>();
+    this.lastNodePerQueue = new VeniceConcurrentHashMap<>(producerThreadCount);
+    this.producingTopic = producingTopic;
+    this.producers = new ArrayList<>(producerThreadCount);
+    this.drainers = new ArrayList<>(producerThreadCount);
+    for (int cur = 0; cur < producerThreadCount; ++cur) {
+      PubSubProducerAdapter producer = producerAdapterSupplier.get();
+      this.producers.add(producer);
+
+      MemoryBoundBlockingQueue<ProducerQueueNode> blockingQueue =
+          new MemoryBoundBlockingQueue<>(producerQueueSize, 1 * 1024 * 1024l);
+      ProducerQueueDrainer drainer = new ProducerQueueDrainer(blockingQueue, cur, producer);
+      this.blockingQueueList.add(blockingQueue);
+      this.drainers.add(drainer);
+      this.producerExecutor.submit(drainer);
+    }
+  }
+
+  public static class ProducerQueueNode implements Measurable {
+    /**
+     * Considering the overhead of {@link PubSubMessage} and its internal structures.
```

Comment:
Nitpick: Maybe this comment reflects an earlier version of your code, where you held onto a `PubSubMessage` instance, but it seems that this is not the case anymore, and you instead hang on to the internal content of the `PSM`, but not the `PSM` itself. Consider tweaking the text accordingly.

Reply:
Fixed.